crypto/multisig/src/v2.rs

```diff
@@ -0,0 +1,1102 @@
+    /// are the same, and easier to follow. The example above simply shows a
+    /// poorly specified signer set.
+    pub fn verify<SIG>(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<SIG>,
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        SIG: Clone + Default + Digestible + Eq + Hash + Message + Ord + Serialize + Signature,
+        S: Verifier<SIG>,
+    {
+        if multi_sig.signatures().len() > MAX_SIGNATURES {
+            return Err(SignatureError::new());
+        }
+
+        // Sort and dedup the list of signatures. Even though we match signers to
+        // signatures (and not the other way around), we still deuplicate the
+        // list of signatures since it is cheap to do, and a reasonable defensive
+        // programming measure.
+        let mut signatures = multi_sig.signatures().to_vec();
+        signatures.sort();
+        signatures.dedup();
+
+        // Verify signatures.
+        let mut signer_identities = self.verify_helper(message, &signatures)?;
+
+        // Sort and dedup the list of matched signer identities.
+        signer_identities.sort();
+        signer_identities.dedup();
+        Ok(signer_identities)
+    }
+
+    // The code that does the actual signature verification. We separate it from the
+    // public `verify` method so that we do not keep re-sorting an already
+    // sorted signers/signatures list.
+    fn verify_helper<SIG>(
+        &self,
+        message: &[u8],
+        signatures: &[SIG],
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        SIG: Clone + Default + Digestible + Eq + Hash + Message + Ord + Serialize + Signature,
+        S: Verifier<SIG>,
+    {
+        // Sort and dedup the list of signers.
+        // We shouldn't be handed a signer set that contains the same signer multiple
+        // times, but just in case someone did do this, this will protect us
+        // from that.
+        let mut signers = self.signers.clone();
+        signers.sort();
+        signers.dedup();
+
+        // See which signers are satisfied by which signatures.
+        let mut matched_signer_identities = Vec::new();
+        let mut num_matched_signers = 0;
+
+        for signer in signers.iter() {
+            match signer.entity {
+                Some(SignerEntity::Single(ref single_signer)) => {
+                    // See if any of the signatures match this signer.
+                    // Note that we do not need to check if we already encounted this signer, since
+                    // we de-duped the list of signers before entering the outer loop.
+                    if signatures
+                        .iter()
+                        .any(|sig| single_signer.verify(message, sig).is_ok())
+                    {
+                        matched_signer_identities.push(single_signer.clone());
+                        num_matched_signers += 1;
+                    }
+                }
```

Comment:
:thought_balloon: Thinking about this some more, this nested looping means that this is On^2 over "single_signer.verify()", which is a hashing function. Worst case I think it's:
```
n * (n + 1)
-----------
      2
```

Would it be better to compute the signature for each signer first (n hashes) and then search for those signatures in the sorted signatures list (log n)? 

I know we're limiting to 10 up above so it may not matter too much.  It would just change the complexity to `n log(n)` where the n part is over more costly hashing.

Reply:
> Would it be better to compute the signature for each signer first (n hashes) and then search for those signatures in the sorted signatures list (log n)?

The signatures here are actually like ed25519 signatures -- the signer is a public key, and the signature can only be created if you know the corresponding private key. The verifier is not in a position to generate the signatures corresponding to a given signer identity.

I think if you want to reduce the algorithmic complexity, a good way to go is to require all the signatures to be labelled in some way with the signers they represent. For instance like, if you have to pair each signature with a copy of the corresponding `SignerIdentity` then this part would become just search over the Signer identities instead of n^2 verify calls. But I'm not sure it's important to optimize this in this way right now.

Reply:
Ok I wouldn't worry about it then, seems like that's just opening the door to false negatives in the verify if someone gets the label logic wrong